### PR TITLE
Use partial message update for pinMessage

### PR DESCRIPTION
**Pull request by AnatolyRugalev**
This PR switches pinMessage to use `partialUpdateMessage` and also adds support for server-side message pinning

### CODE REVIEW

File: src/client.ts [hunk 1]
```diff
@@ -2535,32 +2529,54 @@ export class StreamChat<
...
-    });
+    let messageID: string;
+    if (typeof messageOrMessageID === 'string') {
+      messageID = messageOrMessageID;
+    } else {
+      if (!messageOrMessageID.id) {
+        throw Error('Please specify the message id when calling unpinMessage');
+      }
+      messageID = messageOrMessageID.id;
+    }
```
Comment (ferhatelmas) [inline src/client.ts:2570]: might be better to extract to a helper

**Review by ferhatelmas — APPROVED** [review]
LGTM

File: src/client.ts (hunk 1 above)
Comment (AnatolyRugalev) [reply src/client.ts]: Done

**Review by ferhatelmas — DISMISSED** [review]
LGTM

File: src/client.ts [hunk 2]
```diff
@@ -2535,32 +2523,68 @@ export class StreamChat<
...
-    });
+    return pinExpires;
+  }
+
+  /**
+   * _messageId - extracts string message id from either message object or message id
+   * @param {string | { id: string }} messageOrMessageId message object or message id
+   * @param {string} caller caller function name to report an error in case of message id absence
+   */
+  _messageId(messageOrMessageId: string | { id: string }, caller: string) {
```
Comment (vishalnarkhede) [inline src/client.ts:2534]: how about 
```
_validateAndGetMessageId(messageOrMessageId, "Please specify the message id when calling pinMessage")
```

File: src/client.ts [hunk 3]
```diff
@@ -2535,32 +2523,68 @@ export class StreamChat<
...
+    return this.partialUpdateMessage(
+      this._messageId(messageOrMessageId, 'pinMessage'),
+      {
+        set: {
+          pinned: true,
+          pin_expires: this._normalizeExpiration(timeoutOrExpirationDate),
+        },
+      },
+      userId,
+    );
```
Comment (vishalnarkhede) [inline src/client.ts:2567]:
```suggestion
    const messageId = this._validateAndGetMessageId(
        messageOrMessageId,
        "Please specify the message id when calling pinMessage"
    );

    return this.partialUpdateMessage(
      messageId,
      {
        set: {
          pinned: true,
          pin_expires: this._normalizeExpiration(timeoutOrExpirationDate),
        },
      },
      userId,
    );
```

File: src/client.ts (hunk 2 above)
Comment (vishalnarkhede) [reply src/client.ts:2534]: `_messageId()` doesn't really say much about the function!!

Comment (AnatolyRugalev) [reply src/client.ts:2534]: Agreed, going to change that

File: src/client.ts [hunk 4]
```diff
@@ -2541,32 +2529,76 @@ export class StreamChat<
...
-    });
+    return pinExpires;
+  }
+
+  /**
+   * _messageId - extracts string message id from either message object or message id
+   * @param {string | { id: string }} messageOrMessageId message object or message id
+   * @param {string} message error message to report in case of message id absence
+   */
+  _validateAndGetMessageId(messageOrMessageId: string | { id: string }, message: string) {
```
Comment (vishalnarkhede) [inline src/client.ts:2540]:
```suggestion
  _validateAndGetMessageId(messageOrMessageId: string | { id: string }, errorText: string) {
```

File: src/client.ts [hunk 5]
```diff
@@ -2541,32 +2529,76 @@ export class StreamChat<
...
-      pinned: true,
-      pin_expires: pinExpires,
-    });
+    return pinExpires;
+  }
+
+  /**
+   * _messageId - extracts string message id from either message object or message id
+   * @param {string | { id: string }} messageOrMessageId message object or message id
+   * @param {string} message error message to report in case of message id absence
```
Comment (vishalnarkhede) [inline src/client.ts:2538]:
```suggestion
   * @param {string} errorText error message to report in case of message id absence
```

**Review by vishalnarkhede — APPROVED** [review]

**Pull request merged**